hir2/src/ir/operation/name.rs

```diff
@@ -0,0 +1,191 @@
+///
+/// The operation name consists of its dialect name, and the opcode name within the dialect.
+///
+/// No two operation names can share the same fully-qualified operation name.
+#[derive(Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct OperationName(Rc<OperationInfo>);
+
+struct OperationInfo {
+    /// The dialect of this operation
+    dialect: DialectName,
+    /// The opcode name for this operation
+    name: interner::Symbol,
+    /// The type id of the concrete type that implements this operation
+    type_id: TypeId,
+    /// Details of the traits implemented by this operation, used to answer questions about what
+    /// traits are implemented, as well as reconstruct `&dyn Trait` references given a pointer to
+    /// the data of a specific operation instance.
+    traits: Box<[TraitInfo]>,
+}
+
+impl OperationName {
+    pub fn new<O, S, T>(dialect: DialectName, name: S, traits: T) -> Self
+    where
+        O: crate::Op,
+        S: Into<interner::Symbol>,
+        T: IntoIterator<Item = TraitInfo>,
+    {
+        let type_id = TypeId::of::<O>();
+        let mut traits = traits.into_iter().collect::<Vec<_>>();
+        traits.sort_by_key(|ti| *ti.type_id());
+        let traits = traits.into_boxed_slice();
+        let info = Rc::new(OperationInfo::new(dialect, name.into(), type_id, traits));
+        Self(info)
+    }
+
+    /// Returns the dialect name of this operation
+    pub fn dialect(&self) -> DialectName {
+        self.0.dialect
+    }
+
+    /// Returns the namespace to which this operation name belongs (i.e. dialect name)
+    pub fn namespace(&self) -> interner::Symbol {
+        self.0.dialect.as_symbol()
+    }
+
+    /// Returns the name/opcode of this operation
+    pub fn name(&self) -> interner::Symbol {
+        self.0.name
+    }
+
+    /// Returns true if `T` is the concrete type that implements this operation
+    pub fn is<T: Op>(&self) -> bool {
+        TypeId::of::<T>() == self.0.type_id
+    }
+
+    /// Returns true if this operation implements `Trait`
+    pub fn implements<Trait>(&self) -> bool
+    where
+        Trait: ?Sized + Pointee<Metadata = DynMetadata<Trait>> + 'static,
+    {
+        let type_id = TypeId::of::<Trait>();
+        self.0.traits.binary_search_by(|ti| ti.type_id().cmp(&type_id)).is_ok()
+    }
+
+    /// Returns true if this operation implements `trait`, where `trait` is the `TypeId` of a
+    /// `dyn Trait` type.
+    pub fn implements_trait_id(&self, trait_id: &TypeId) -> bool {
+        self.0.traits.binary_search_by(|ti| ti.type_id().cmp(trait_id)).is_ok()
+    }
+
+    #[inline]
+    pub(super) fn downcast_ref<T: Op>(&self, ptr: *const ()) -> Option<&T> {
+        if self.is::<T>() {
+            Some(unsafe { self.downcast_ref_unchecked(ptr) })
+        } else {
+            None
+        }
+    }
+
+    #[inline(always)]
+    unsafe fn downcast_ref_unchecked<T: Op>(&self, ptr: *const ()) -> &T {
```

Comment:
I'm a fan of op traits implementation! TIL about `DynMetadata`. So cool!

Reply:
I'm definitely happy with how that has turned out so far, it enables a number of really neat tricks! You might also be interested in a bit of trivia about pointer metadata in Rust:

* `DynMetadata<dyn Trait>` is the type of metadata associated with a trait object for the trait `Trait`, and is basically just a newtype for a pointer to the vtable of the trait object. While you could specify `DynMetadata<T>` for some `T` that _isn't_ a trait, that doesn't actually mean anything, and is technically invalid AFAIK.
* `usize` is  the type of metadata associated with "unsized" types (except trait objects), and corresponds to the unit size for the pointee type. For example, `&[T]` is an unsized reference, and its pointer metadata is the number of `T` in the slice, i.e. a typical "fat" pointer; but the semantics of the metadata value depend on the type, so you can have custom types that use `usize` metadata to indicate the allocation size of the pointee, or something else along those lines.
* `()` is the type of metadata associated with all "sized" types, i.e. pointers to those types don't actually have metadata

Reply:
Thank you! I'm definitely going to dig into it.